Hollweg dispersion solver

Sjbrownian opened the pull request:
<!--
Thank you for contributing to PlasmaPy! Here's a bunch of pointers to
make things easier for all of us:

* If this PR will solve an issue tracked by GitHub, then please add
  "Closes #42" so the issue automatically closes once this pull request
  is merged.  In this example, issue #42 would have been closed.
  If your PR will not completely solve the issue, then please still reference
  still reference the issue like "issue #42".  (For more info see [GitHub's primer
  on linking issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).)

* Remember to add some description of your changes in this text box.

* If your pull request is not yet ready for review - either because
  you're just looking for feedback on a change or because you're not
  perfectly satisfied with your change - submit it as a draft pull request.
  Remember to change its' status once it's ready.

* If this is your first contribution, please add your name to the author
  list in `docs/about/credits.rst`.

* Feel free to chat with other developers on our Matrix channel at:
   https://app.element.io/#/room/#plasmapy:openastronomy.org

* We have a developer's guide to help answer some of your questions.
  http://docs.plasmapy.org/en/latest/development/index.html

Many thanks in advance for following these pointers and for being willing to contribute!

When submitting a pull request, please ensure that you can (eventually,
sometime before it is merged) check the following basic requirements:

-->

- [ ] I have added a changelog entry for this pull request.

<!--

In short: A changelog entry is a short description of your PR's changes.
Each entry is written in a `<PULL REQUEST>.<TYPE>.rst` file and stored in
the `changelog` directory,  where `<PULL REQUEST>` is a pull request
number and `<TYPE>` is one of:

* `breaking`: A change which requires users to change code and is not backwards compatible. (Not to be used for removal of deprecated features.)
* `feature`: New user facing features and any new behavior.
* `bugfix`: Fixes a reported bug.
* `doc`: Documentation addition or improvement, like rewording an entire session or adding missing docs.
* `removal`: Feature deprecation and/or feature removal.
* `trivial`: A change which has no user facing effect or is tiny change.

A PR number is generated after you successfully submit a new PR, so the changelog
file can only be added after you open the PR.```

For more information, see:
https://github.com/PlasmaPy/PlasmaPy/blob/master/changelog/README.rst

--->

- [ ] If adding new functionality, I have added tests and
      docstrings.

<!--
(Tests pop up at the bottom, in the checks box.)
-->

- [ ] I have fixed any newly failing tests.

<!--
(If you're unsure why they're failing, ask!)
-->

rocco8773 commented:
@Sjbrownian now that PR #1208 is merged go ahead and update this branch with the updated `main`.  Once you do that and resolve any of the conflicts I can start looking over the functionality.

Sjbrownian commented:
> @Sjbrownian now that PR #1208 is merged go ahead and update this branch with the updated `main`. Once you do that and resolve any of the conflicts I can start looking over the functionality.

@rocco8773 I've updated the branch and fixed the code so that most of the tests pass. It's not liking the way the docstring is formatted for some reason which makes a couple tests fail. The changelog test fails but I'm not sure if it's urgent or if it's something that we can wait on. Also the `pre-commit.ci - pr` test is failing but I'm a little fuzzy on what we did to fix that issue with the last PR.
Are enough of the conflicts resolved for you to start reviewing the functionality or is there anything else I should do?

rocco8773 commented:
Yeah, that should be good enough for me to get started. The current check failures or only style checks, so there's no urgency. I'll point out the fixes as I do the review.

rocco8773 commented on plasmapy/dispersion/numerical/hollweg_.py:
```diff
@@ -0,0 +1,204 @@
+import numpy as np
+import astropy.units as u
+import warnings
+
+from astropy.constants.si import c
+from plasmapy.formulary import parameters as pfp
+from plasmapy.particles import Particle
+from plasmapy.particles.exceptions import ChargeError
+from plasmapy.utils.decorators import validate_quantities
+from plasmapy.utils.exceptions import PhysicsWarning
+
+from typing import Union
```
This is one of the things the `pre-commit` check is complaining about.  The import order needs to be done in a specific/consistent way.  There rules for this are defined in the `[tool.isort]` section in the `pyproject.toml` file.  You can always check this yourself locally by using the `isort` package.  So, ...

1. make sure you have `isort==5.7.0` installed (`pip install isort==5.7.0`)
2. once installed, open a terminal and navigate to the root directory of the repo
3. run the following command

    ```bash
    python -m isort --diff --color plasmapy
    ```

   This will run `isort` on the whole `plasmapy` pkg.  If you want to just run it on a specific file then supply that path instead of just `plasmapy`.  The `--diff` option means it will only show you what to change.  If you exclude `--diff` then it'll make the changes for you.  The `--color` option just means it will color the output for (easier to read).

rocco8773 commented on plasmapy/dispersion/numerical/__init__.py:
```diff
@@ -2,4 +2,7 @@
 associated with numerical dispersion solvers.
 """
-__all__ = []
+__all__ = [
+  "hollweg",
+]
```
```suggestion
__all__ = ["hollweg"]

```

This does not need to be split onto multiple lines until the code reaches the max line limit.  When to start the multiple line split is a little subjective, but it's based on readability and I usually don't split the list until it gets over 3 elements, or it reaches the max line length.

rocco8773 commented on docs/notebooks/dispersion/hollweg_dispersion.ipynb:
```diff
@@ -0,0 +1 @@
+
```
Delete this until you actually start the associated notebook.

rocco8773 reviewed (changes requested):
Similar to `isort`, we use the `black` package to check that code is formatted properly, and there appears to be some format issues in `hollweg_.py`.  Like `isort` do..

1. ensure `black==21.4b2` is installed
2. navigate to the root directory of `plasmapy`
3. run the following command (essentially same arguments as `isort`)
```python
python -m black --diff --color plasmapy
```

Sjbrownian commented:
> Similar to `isort`, we use the `black` package to check that code is formatted properly, and there appears to be some format issues in `hollweg_.py`. Like `isort` do..
> 
> 1. ensure `black==21.4b2` is installed
> 2. navigate to the root directory of `plasmapy`
> 3. run the following command (essentially same arguments as `isort`)
> 
> ```python
> python -m black --diff --color plasmapy
> ```

@rocco8773 Thank you! I had the commands written down from the office hours but I was under the weather at the time and forgot how we applied them. This helped!. `isort` didn't give me any errors when I ran it but `black` showed me all the formatting issues and the `pre-commit - pr` check is passing now.

rocco8773 commented on plasmapy/dispersion/numerical/hollweg_.py:
```diff
@@ -0,0 +1,271 @@
+    ... }
+    >>> omegas = hollweg(**inputs)
+    >>> omegas
+    {'fast_mode': <Quantity [2.62911663e-02, 2.27876968e+03] rad / s>,
+    'alfven_mode': <Quantity [7.48765909e-04, 2.13800404e+03] rad / s>,
+    'acoustic_mode': <Quantity [0.00043295, 0.07358991] rad / s>}"""
```
```suggestion
    {'fast_mode': <Quantity [2.62911663e-02, 2.27876968e+03] rad / s>,
     'alfven_mode': <Quantity [7.48765909e-04, 2.13800404e+03] rad / s>,
     'acoustic_mode': <Quantity [0.00043295, 0.07358991] rad / s>}
    """
```

I believe this will get the "Linter" check to pass.

Sjbrownian replied:
I think this was one of the things setting the test off. There were some other spacing issues that it didn't like but the test passes now.

rocco8773 reviewed (changes requested):
Looks like you figured out most of the documentation issues. Nice!  I was getting ready to add them to a review and then I got sucked into meetings.

I have two more documentation items to add:

1. The `numerical` sub-package needs to be imported into `plasmapy/dispersion/__init__.py`.  So the imports should look something like...

```python
__all__ = [
    "plasma_dispersion_func",
    "plasma_dispersion_func_deriv",
]

from plasmapy.dispersion import analytical, numerical
from plasmapy.dispersion.dispersionfunction import (
    plasma_dispersion_func,
    plasma_dispersion_func_deriv,
)
```

2. A stub file needs to be created for the `numerical` sub-package.  This is done by creating and adding the file `plasmapy.dispersion.numerical.rst` to the `docs/api_static` directory.  The contents of this file should look like...

```rest
:orphan:

`plasmapy.dispersion.numerical`
===============================

.. currentmodule:: plasmapy.dispersion.numerical

.. automodapi:: plasmapy.dispersion.numerical

```

rocco8773 commented on plasmapy/dispersion/numerical/hollweg_.py:
```diff
@@ -0,0 +1,285 @@
+    -----
+    Solves the equation 3 in Bellan2012JGR (equation 38 in Hollweg1999)
+
+    .. math::
+        \left(\frac{\omega^2}{k_{\rm z}^2 v_{\rm A}^2} - 1 \right) &
+        \left[\omega^2\left(\omega^2 - k^2 v_{\rm A}^2\right) - &
+        \beta k^2 v_{\rm A}^2 \left(\omega^2 - k_{\rm z}^2 v_{\rm A}^2 \right) &
+        \right]\\ = \omega^2 \left(\omega^2 - k^2 v_{\rm A}^2 \right) k_{\rm x}^2 &
+        \left(\frac{c_{\rm s}^2}{\omega_{\rm ci}^2} - \frac{c^2}{\omega_{\rm pe}^2} &
+        \frac{\omega^2}{k_{\rm z}^2v_{\rm A}^2}\right)
```
```suggestion
    .. math::
        \left( \frac{\omega^2}{k_{\rm z}^2 v_{\rm A}^2} - 1 \right) &
        \left[
            \omega^2 \left( \omega^2 - k^2 v_{\rm A}^2 \right)
            - \beta k^2 v_{\rm A}^2 \left(
                \omega^2 - k_{\rm z}^2 v_{\rm A}^2
            \right)
        \right] \\
        &= \omega^2 \left(\omega^2 - k^2 v_{\rm A}^2 \right) k_{\rm x}^2
        \left(
            \frac{c_{\rm s}^2}{\omega_{\rm ci}^2}
            - \frac{c^2}{\omega_{\rm pe}^2} \frac{\omega^2}{k_{\rm z}^2v_{\rm A}^2}
        \right)
```

This should get the math to render correctly in the documentation build.

Sjbrownian replied:
Is there a way to check what the math render looks like as I type out the text so I can check if it's correct when I do this in the future?

rocco8773 replied:
You essentially have two options:

1. Write it up in a Latex editor and then copy-paste.
2. Build a local html copy of the documentation and launch it your web browser.  Nick did do a write up on this in the [documentation guide](https://docs.plasmapy.org/en/latest/development/doc_guide.html#building-documentation).  In his write up he covers using `make` and `tox` to build documentation, but, if I remember correctly, you're using windows and the `make` command does not work properly in the command prompt terminal or powershell.  To use `make` you'd need to setup and use the [wsl terminal](https://docs.microsoft.com/en-us/windows/wsl/about) (this is what I do).

If we have time, I can elaborate on this during the dispersion meeting this week.

rocco8773 commented on plasmapy/dispersion/numerical/hollweg_.py:
```diff
@@ -0,0 +1,285 @@
+import astropy.units as u
```
Add a short module level docstring at the top of the file.

Sjbrownian commented:
@rocco8773 I've added a short (in progress) description at the start of the docstring, a little bit at the start of the Notes section, and I've added the References. The check for building the documentation is failing for some reason. I've tried making some changes and reading the errors the check is giving me but haven't had any luck. Would you be able to take a look when you get a chance?

rocco8773 commented:
@Sjbrownian I am unsure what is causing the RTD build to fail, but it does not appear to be related to something you did.  I'll have to investigate further and let you know, but don't worry about it for now.

rocco8773 commented:
@Sjbrownian This seemed to be an issue related to `numpy`.  `numpy` has a package version and C API version.  Somehow these two API versions got out of sync on RTD which caused it to fail.  .... I don't completely understand it myself, but it's working now and I'll take it.

I will make this change official through a separate PR, so don't worry about making a change log entry for it.

Sjbrownian commented:
> @Sjbrownian This seemed to be an issue related to `numpy`. `numpy` has a package version and C API version. Somehow these two API versions got out of sync on RTD which caused it to fail. .... I don't completely understand it myself, but it's working now and I'll take it.
> 
> I will make this change official through a separate PR, so don't worry about making a change log entry for it.

Sounds good, thank you.

Sjbrownian commented:
@rocco8773 I've added references for the Hollweg and Bellan papers, and I added the assumptions made to derive the dispersion relation in the docstring. Is there anything else you need me to add before you start reviewing the functionality?

rocco8773 commented:
I'll read over the docstrings and make any suggestions if needed.  I do think you're at a point now where you can remove the Draft mark if you feel it's appropriate.

Sjbrownian commented:
> I'll read over the docstrings and make any suggestions if needed. I do think you're at a point now where you can remove the Draft mark if you feel it's appropriate.

Sounds good, I'll remove the draft mark.

rocco8773 commented on plasmapy/dispersion/numerical/hollweg_.py:
```diff
@@ -0,0 +1,321 @@
+    alpha = (k * v_A) ** 2
+
+    # Polynomial coefficients: c3*x^3 + c2*x^2 + c1*x + c0 = 0
+    c3 = (F * kx ** 2 + 1) / sigma
+    c2 = -((alpha / sigma) * (1 + beta + F * kx ** 2) + D * kx ** 2 + 1)
+    c1 = alpha * (1 + 2 * beta + D * kx ** 2)
+    c0 = -beta * alpha * sigma
```
I went through the math and was able to derive the same expressions for these coefficients.  However, this formulation will not play nicely with the edge case of `kz == 0` (i.e. `sigma == 0`).  Not sure what the best approach is to remedy this, but most likely just multiple the expressions by `sigma` to eliminate the divide by zero.

Sjbrownian replied:
I forgot to consider the divide by 0 case. I multiplied everything by sigma and double checked that the output was the same just be thorough. This should be fixed now.

rocco8773 commented on plasmapy/dispersion/numerical/hollweg_.py:
```diff
@@ -0,0 +1,338 @@
+    T_i : `~astropy.units.Quantity`
+        The ion temperature in units of K or eV.
+    theta : `~astropy.units.Quantity`, single valued or 1-D array
+        The angle of propagation of the wave with respect to the magnetic
+        field, :math:`\cos^{-1}(k_z / k)`, in units convertible to
+        radians.  Either single valued or 1-D array of size :math:`M`.
```
```suggestion
        The angle of propagation of the wave with respect to the
        magnetic field, :math:`\cos^{-1}(k_z / k)`, in units convertible
        to radians.  Either single valued or 1-D array of size
        :math:`M`.
```

rocco8773 commented on plasmapy/dispersion/numerical/hollweg_.py:
```diff
@@ -0,0 +1,339 @@
+    # validate z_mean
+    if z_mean is None:
+        try:
+            z_mean = abs(ion.integer_charge)
```
```suggestion
            z_mean = abs(ion.charge_number)
```

name `integer_charge` as dropped for `charge_number` in PR #1136

rocco8773 commented on plasmapy/dispersion/numerical/hollweg_.py:
```diff
@@ -0,0 +1,339 @@
+        k = np.array([k.value,]) * u.rad / u.m
+
+    # Calc needed plasma parameters
+    n_e = z_mean * n_i
+    c_s = pfp.ion_sound_speed(
+        T_e=T_e,
+        T_i=T_i,
+        ion=ion,
+        n_e=n_e,
+        gamma_e=gamma_e,
+        gamma_i=gamma_i,
+        z_mean=z_mean,
+    )
+    v_A = pfp.Alfven_speed(B, n_i, ion=ion, z_mean=z_mean)
+    omega_ci = pfp.gyrofrequency(B=B, particle=ion, signed=False, Z=z_mean)
+    omega_pe = pfp.plasma_frequency(n=n_e, particle="e-")
```
```suggestion
    with warnings.catch_warnings():
        warnings.simplefilter("ignore", category=PhysicsWarning)
        n_e = z_mean * n_i
        c_s = pfp.ion_sound_speed(
            T_e=T_e,
            T_i=T_i,
            ion=ion,
            n_e=n_e,
            gamma_e=gamma_e,
            gamma_i=gamma_i,
            z_mean=z_mean,
        ).value
        v_A = pfp.Alfven_speed(B, n_i, ion=ion, z_mean=z_mean).value
        omega_ci = pfp.gyrofrequency(B=B, particle=ion, signed=False, Z=z_mean).value
        omega_pe = pfp.plasma_frequency(n=n_e, particle="e-").value

    # strip units from select input args
    k = k.value
    theta = theta.value
```

Filtering out the `PhysicsWarning` issued by `ion_sound_speed` since this is the setup we're expecting.

Dropping all astropy units in preparation for calculation later on.

rocco8773 commented on plasmapy/dispersion/numerical/hollweg_.py:
```diff
@@ -0,0 +1,339 @@
+    cs_vA = c_s / v_A
+
+    # Parameters kx and kz
+    kz = np.cos(theta.value) * k
+    kx = np.sqrt(k ** 2 - kz ** 2)
```
```suggestion
    kz = np.cos(theta) * k
    kx = np.sin(theta) * k
```

We have already stripped units so we do not have to explicitly do `theta.value`.

It's up to you to use sine of theta or the method you originally wrote.  Since all the later calculations use `kx ** 2`, the sign should not matter.

rocco8773 commented on plasmapy/dispersion/numerical/hollweg_.py:
```diff
@@ -0,0 +1,339 @@
+    kz = np.cos(theta.value) * k
+    kx = np.sqrt(k ** 2 - kz ** 2)
+
+    # Bellan2012JGR beta param equation 3
+    beta = (c_s / v_A) ** 2
+
+    # Parameters D, F, sigma, and alpha to simplify equation 3
+    D = (c_s / omega_ci) ** 2
+    F = (c / omega_pe) ** 2
+    sigma = (kz * v_A) ** 2
+    alpha = (k * v_A) ** 2
```
```suggestion
    # Define helpful parameters
    beta = (c_s / v_A) ** 2
    alpha_A = (k * v_A) ** 2
    alpha_s = (k * c_s) ** 2  # == alpha_A * beta
    sigma = (kz * v_A) ** 2
    D = (c_s / omega_ci) ** 2
    F = (c_si_unitless / omega_pe) ** 2
```

I introduced `alpha_s` instead of multiply `alpha_A * beta` since the later would involve extra calculation of multiplying by `v_A` and then dividing by `v_A`

Sjbrownian replied:
@rocco8773 When defining `F`, the `c_si_unitless` is giving me an error. Should this just be returned to `c` or should something else replace it?

Sjbrownian replied:
I changed it to `c.value` for the moment, that seems to work, let me know if you had a different intention here though.

rocco8773 commented on plasmapy/dispersion/numerical/hollweg_.py:
```diff
@@ -0,0 +1,339 @@
+    alpha = (k * v_A) ** 2
+
+    # Polynomial coefficients: c3*x^3 + c2*x^2 + c1*x + c0 = 0
+    c3 = F * kx ** 2 + 1
+    c2 = -sigma * ((alpha / sigma) * (1 + beta + F * kx ** 2) + D * kx ** 2 + 1)
+    c1 = sigma * alpha * (1 + 2 * beta + D * kx ** 2)
+    c0 = -beta * alpha * sigma ** 2
```
```suggestion
    c3 = F * kx ** 2 + 1
    c2 = -alpha_A * (1 + beta + F * kx ** 2) - sigma * (1 + D * kx ** 2)
    c1 = sigma * alpha_A * (1 + 2 * beta + D * kx ** 2)
    c0 = -alpha_s * sigma ** 2
```

Applying changes from https://github.com/PlasmaPy/PlasmaPy/pull/1189/files#r767014478

rocco8773 commented on plasmapy/dispersion/numerical/hollweg_.py:
```diff
@@ -0,0 +1,339 @@
+    c1 = sigma * alpha * (1 + 2 * beta + D * kx ** 2)
+    c0 = -beta * alpha * sigma ** 2
+
+    omega = {}
+    fast_mode = []
+    alfven_mode = []
+    acoustic_mode = []
+
+    # a3*x^3 + a2*x^2 + a1*x + a0 = 0
+    for (a3, a2, a1, a0) in zip(c3, c2, c1, c0):
+
+        w = np.emath.sqrt(np.roots([a3.value, a2.value, a1.value, a0.value]))
+        fast_mode.append(np.max(w))
+        alfven_mode.append(np.median(w))
+        acoustic_mode.append(np.min(w))
+
+    omega["fast_mode"] = fast_mode * u.rad / u.s
+    omega["alfven_mode"] = alfven_mode * u.rad / u.s
+    omega["acoustic_mode"] = acoustic_mode * u.rad / u.s
+
+    omega["fast_mode"] = omega["fast_mode"].squeeze()
+    omega["alfven_mode"] = omega["alfven_mode"].squeeze()
+    omega["acoustic_mode"] = omega["acoustic_mode"].squeeze()
+
+    # check the low-frequency limit
+
+    m1 = np.max(omega["fast_mode"])
+    m2 = np.max(omega["alfven_mode"])
+    m3 = np.max(omega["acoustic_mode"])
```
```suggestion
    # Find roots to polynomial
    coefficients = np.array([c3, c2, c1, c0], ndmin=2)
    nroots = coefficients.shape[0] - 1  # 3
    nks = coefficients.shape[1]
    roots = np.empty((nroots, nks), dtype=np.complex128)
    for ii in range(nks):
        roots[:, ii] = np.roots(coefficients[:, ii])[:]

    roots = np.sqrt(roots)
    roots = np.sort(roots, axis=0)
```

rocco8773 commented on plasmapy/dispersion/numerical/hollweg_.py:
```diff
@@ -0,0 +1,339 @@
+    m2 = np.max(omega["alfven_mode"])
+    m3 = np.max(omega["acoustic_mode"])
+
+    w_max = max(m1, m2, m3)
+    w_wci_max = w_max / omega_ci
+
+    # Warnings
+
+    if c_s / v_A > 0.1:
+        warnings.warn(
+            f" This solver is valid in the regime c_s/v_A << 1. "
+            f"A c_s/v_A value of {cs_vA:.2f} was calculated which "
+            f"may affect the validity of the solution.",
+            PhysicsWarning,
+        )
+
+    # dispersion relation is only valid in the regime w << w_ci
+    if w_max / omega_ci > 0.1:
+        warnings.warn(
+            f"This solver is valid in the regime w/w_ci << 1. "
+            f"A w value of {w_max:.2f} and a w/w_ci value of {w_wci_max:.2f} "
+            f"were calculated which may affect the validity of the solution.",
+            PhysicsWarning,
+        )
```
```suggestion
    # Warn about NOT low-beta
    if c_s / v_A > 0.1:
        warnings.warn(
            f"This solver is valid in the low-beta regime, "
            f"c_s/v_A << 1. A c_s/v_A value of {cs_vA:.2f} was"
            f"calculated which may affect the validity of the "
            f"solution.",
            PhysicsWarning,
        )

    # dispersion relation is only valid in the regime w << w_ci
    w_max = np.max(roots)
    w_wci_max = w_max / omega_ci
    if w_WCI_MAX > 0.1:
        warnings.warn(
            f"This solver is valid in the regime w/w_ci << 1.  A w "
            f"value of {w_max:.2f} and a w/w_ci value of "
            f"{w_wci_max:.2f} were calculated which may affect the "
            f"validity of the solution.",
            PhysicsWarning,
        )
```

rocco8773 replied:
As mention in today's dispersion meeting, the warning message for low-beta should still reference Sec.1.7 of Bellan 2012 and prompt users to the plasmapy documentation for full citation.

rocco8773 commented on plasmapy/dispersion/numerical/hollweg_.py:
```diff
@@ -0,0 +1,339 @@
+            PhysicsWarning,
+        )
+
+    return omega
```
```suggestion
    omegas = {
        "acoustic_mode": roots[0, :].squeeze() * u.rad / u.s,
        "alfven_mode": roots[1, :].squeeze()  * u.rad / u.s,
        "fast_mode": roots[2, :].squeeze()  * u.rad / u.s,
    }
    return omegas
```

rocco8773 reviewed (changes requested):
As we discuss earlier today, your code is looking good and these suggestions are to consolidate the calculations so they utilize more `numpy` math.

I did test all these changes on my local copy and they all pass your current tests.

rocco8773 commented on plasmapy/dispersion/numerical/hollweg_.py:
```diff
@@ -0,0 +1,352 @@
+from plasmapy.particles import Particle
+from plasmapy.particles.exceptions import ChargeError
+from plasmapy.utils.decorators import validate_quantities
+from plasmapy.utils.exceptions import PhysicsWarning
+
+
+@validate_quantities(
+    B={"can_be_negative": False},
+    n_i={"can_be_negative": False},
+    T_e={"can_be_negative": False, "equivalencies": u.temperature_energy()},
+    T_i={"can_be_negative": False, "equivalencies": u.temperature_energy()},
+)
```
```suggestion
from plasmapy.utils.exceptions import PhysicsWarning

c_si_unitless = c.value


@validate_quantities(
    B={"can_be_negative": False},
    n_i={"can_be_negative": False},
    T_e={"can_be_negative": False, "equivalencies": u.temperature_energy()},
    T_i={"can_be_negative": False, "equivalencies": u.temperature_energy()},
)
```

rocco8773 commented on plasmapy/dispersion/numerical/hollweg_.py:
```diff
@@ -0,0 +1,352 @@
+    alpha_s = (k * c_s) ** 2  # == alpha_A * beta
+    sigma = (kz * v_A) ** 2
+    D = (c_s / omega_ci) ** 2
+    F = (c.value / omega_pe) ** 2
```
```suggestion
    F = (c_si_unitless / omega_pe) ** 2
```

rocco8773 reviewed (changes requested):
This is a minor thing, but by stripping the units on the module level you don't incur the time penalty on the function level.

rocco8773 commented on plasmapy/dispersion/numerical/hollweg_.py:
```diff
@@ -0,0 +1,352 @@
+        )
+
+    # Single k value case
+    if np.isscalar(k.value):
+        k = (
+            np.array(
+                [
+                    k.value,
+                ]
+            )
+            * u.rad
+            / u.m
+        )
```
```suggestion
    if np.isscalar(k.value):
        k = np.array([k.value]) * u.rad / u.m
```

A cleaner way that does not make `black` upset.  I always forget that having the comma `[k.value,]` causes `black` to want to split everything onto its own line.

rocco8773 commented on plasmapy/dispersion/numerical/hollweg_.py:
```diff
@@ -0,0 +1,352 @@
+        )
+
+    # Warn about theta not nearly perpendicular
+    if theta < 0.5 * np.pi - 0.1 or theta > 0.5 * np.pi + 0.1:
```
```suggestion
    if np.abs(theta - np.pi / 2) > 0.1:
```

Just for improved readability.

rocco8773 commented on plasmapy/dispersion/numerical/hollweg_.py:
```diff
@@ -0,0 +1,352 @@
+        warnings.warn(
+            f"This solver is valid in the regime where propogation is "
+            f"nearly perpendicular to B according to Bellan, 2012, Sec. 1.7 "
+            f"(see documentation for DOI). A theta value of {theta:.2f} was "
+            f" entered which may affect the validity of the solution.",
```
```suggestion
            f"(see documentation for DOI). A theta value of {theta:.2f} was "
            f"entered which may affect the validity of the solution.",
```

Remove extra space so there's not two between the words.

rocco8773 commented on plasmapy/dispersion/numerical/hollweg_.py:
```diff
@@ -0,0 +1,352 @@
+
+    Warns
+    -----
+    : `~plasmapy.utils.exceptions.PhysicsWarning`
+        When the computed wave frequencies violate the
+        :math:`\omega/\omega_{\rm ci} \ll 1` assumption of the
+        dispersion relation.
```
```suggestion
    : `~plasmapy.utils.exceptions.PhysicsWarning`
        When :math:`\omega / \omega_{\rm ci} > 0.1`, violation of the
        low-frequency assumption.

    : `~plasmapy.utils.exceptions.PhysicsWarning`
        When :math:`c_{\rm s} / v_{\rm A} > 0.1`, violation of low-β.

    : `~plasmapy.utils.exceptions.PhysicsWarning`
        When :math:`|θ - π/2| > 0.1`, violation of quasi-perpendicular
        propagation.
```

rocco8773 commented on plasmapy/dispersion/numerical/hollweg_.py:
```diff
@@ -0,0 +1,352 @@
+        `~plasmapy.particles.Particle`.
+
+    TypeError
+        If ``gamma_e``, ``gamma_i``, or``z_mean`` are not of type `int`
```
```suggestion
        If ``gamma_e``, ``gamma_i``, or ``z_mean`` are not of type `int`
```

rocco8773 reviewed (changes requested):
These are updates to the docstring so they're consistent with the updates you've added (primarily the additional assumptions).  A few syntax errors are fixed.  Updated citations to utilized the bibliography implement by PR #1263.

To get the citations to work you'll need to add the following entry to the `/docs/bibliography.bib` file.  Just place the entry alphabetically in the file.

```bibtex
@article{hollweg:1999,
   author = {Hollweg, Joseph. V.},
   title = {Kinetic Alfvén wave revisited},
   journal = {Journal of Geophysical Research},
   volume = {104},
   number = {A7},
   doi = {10.1029/1998JA900132},
   year = {1999}
}
```

Sjbrownian commented:
> These are updates to the docstring so they're consistent with the updates you've added (primarily the additional assumptions). A few syntax errors are fixed. Updated citations to utilized the bibliography implement by PR #1263.
> 
> To get the citations to work you'll need to add the following entry to the `/docs/bibliography.bib` file. Just place the entry alphabetically in the file.
> 
> ```
> @article{hollweg:1999,
>    author = {Hollweg, Joseph. V.},
>    title = {Kinetic Alfvén wave revisited},
>    journal = {Journal of Geophysical Research},
>    volume = {104},
>    number = {A7},
>    doi = {10.1029/1998JA900132},
>    year = {1999}
> }
> ```



> These are updates to the docstring so they're consistent with the updates you've added (primarily the additional assumptions). A few syntax errors are fixed. Updated citations to utilized the bibliography implement by PR #1263.
> 
> To get the citations to work you'll need to add the following entry to the `/docs/bibliography.bib` file. Just place the entry alphabetically in the file.
> 
> ```
> @article{hollweg:1999,
>    author = {Hollweg, Joseph. V.},
>    title = {Kinetic Alfvén wave revisited},
>    journal = {Journal of Geophysical Research},
>    volume = {104},
>    number = {A7},
>    doi = {10.1029/1998JA900132},
>    year = {1999}
> }
> ```

I've added this and committed the rest of the changes you commented. Sorry I took so long to get to this, I've had to focus on finals these past two weeks.

StanczakDominik reviewed (approved):
Looks great! I have a few small suggestions, but all in all, :+1+. Thanks for doing this!

rocco8773 commented on plasmapy/dispersion/numerical/tests/test_hollweg_.py:
```diff
@@ -0,0 +1,172 @@
+            ),
+        ],
+    )
+    def test_return_structure(self, kwargs, expected):
+        """Test the structure of the returned values."""
+        ws = hollweg(**kwargs)
+
+        assert isinstance(ws, dict)
+        assert len({"acoustic_mode", "alfven_mode", "fast_mode"} - set(ws.keys())) == 0
+
+        for mode, val in ws.items():
+            assert isinstance(val, u.Quantity)
+            assert val.unit == u.rad / u.s
+            assert val.shape == expected["shape"]
```
@Sjbrownian is still working on the test that actually tests values.  He copied the test structure from the `two_fluid` tests, so in the end these test will look like those.

Sjbrownian replied:
@rocco8773 I made a test for the graph values and they all pass with atol=1e-2. I don't have anything to verify the kx*L (x-axis graph values) or the 4 different beta values shown on the graph though. I'm unsure what the best way to do this would be, I can think of an ugly way with a bunch of if statements but I'm guessing there's a better method.

Maybe we could meet sometime to discuss? I know its getting close to the Holidays so if afterwards is best for you, that works for me.

StanczakDominik replied:
I've made a comment on the tolerances here: https://github.com/PlasmaPy/PlasmaPy/pull/1189#discussion_r774451097 .

I don't think you really need to check `kx*L` here (by the way, if I've got the paper correct it looks like fig2 has `ky*L`, so I'm assuming a typo here?), as long as the y-axis values fit.

I also see that the test cases you're using are point values for x (or rather `ky*L`) `= {0, 1}`. Just to float the idea: if you wanted to compare to the actual curves in a sense, you could use something like https://markummitchell.github.io/engauge-digitizer/ and parametrize over pairs (value of beta, dataframe with `kx*L` and output values). I see the code takes 1D arrays for `k`, so that could be easy to accomplish.

Sjbrownian replied:
Yeah so these dispersion solvers we're working on are from the Bellan2012 paper (attached here, see eqn 3) which uses kx. The Hollweg1999 paper uses ky, and Bellan summarizes results from several papers so my guess is he just changed everything to kx to be consistant. Everything is the same just a change is coordinate system.

I didn't know of enguage-digitizer, I was using WebPlotDigitizer. I will look into this, thank you for the idea!


[Bellan2012.pdf](https://github.com/PlasmaPy/PlasmaPy/files/7769515/Bellan2012.pdf)

StanczakDominik replied:
Oooh, okay, so I did have the wrong paper. Document, document, document that in the test docstring! :)

Sjbrownian replied:
I should have put the Bellan paper in the docstring, thank you for pointing this out!

rocco8773 commented on plasmapy/dispersion/numerical/hollweg_.py:
```diff
@@ -0,0 +1,360 @@
+     'acoustic_mode': <Quantity [0.00043295+0.j, 0.07358991+0.j] rad / s>}
+    """
+
+    # validate argument ion
+    if not isinstance(ion, Particle):
+        try:
+            ion = Particle(ion)
+        except TypeError:
+            raise TypeError(
+                f"For argument 'ion' expected type {Particle} but got {type(ion)}."
+            )
+    if not (ion.is_ion or ion.is_category("element")):
+        raise ValueError("The particle passed for 'ion' must be an ion or element.")
```
Honestly, I don't remember the exact reason.  It was either `@particle_input` could not direction validate on `is_ion` and/or it didn't work well with the typing `Union[str, Particle]`.

rocco8773 commented on plasmapy/dispersion/numerical/hollweg_.py:
```diff
@@ -0,0 +1,360 @@
+        roots[:, ii] = np.roots(coefficients[:, ii])[:]
+
+    roots = np.sqrt(roots)
+    roots = np.sort(roots, axis=0)
```
To ensure the root magnitudes are properly ordered, I have no rationale for assuming `numpy.roots` will do that.  Later on the array indices are assigned to be the fast, slow, and Alfven modes.

rocco8773 commented on plasmapy/dispersion/numerical/hollweg_.py:
```diff
@@ -0,0 +1,360 @@
+
+    # Find roots to polynomial
+    coefficients = np.array([c3, c2, c1, c0], ndmin=2)
+    nroots = coefficients.shape[0] - 1  # 3
```
ummm....that seems weird to me.  Not used to seeing assert statements outside of tests.

StanczakDominik replied:
I've been just sticking them in a little more recently and they've saved me some debugging time since! Still, it's fine to keep this that way.

rocco8773 commented on plasmapy/dispersion/numerical/hollweg_.py:
```diff
@@ -0,0 +1,360 @@
+    # Define helpful parameters
+    beta = (c_s / v_A) ** 2
+    alpha_A = (k * v_A) ** 2
+    alpha_s = (k * c_s) ** 2  # == alpha_A * beta
```
I was trading accuracy for performance here.  If we really wanted performance then we wouldn't be doing these readability declarations and just write everything out in one line.

The `alpha_A * beta` has a `v_A / v_A` and I just wanted to avoid that extra calculation.

StanczakDominik commented on plasmapy/dispersion/numerical/tests/test_hollweg_.py:
```diff
@@ -119,6 +116,83 @@ def test_warning(self, kwargs, _warning):
+        w_alfven = (hollweg(**kwargs)["alfven_mode"]).value
+        big_omega = np.abs(w_alfven / (kz * va))
+
+        assert np.isclose(big_omega, expected, atol=1e-2)
```
I swear half of my maintainer life is just replacing `assert np.isclose` with
```suggestion
        np.testing.assert_allclose(big_omega, expected, rtol=1e-2)
```
in every test for its superior debugging output, lol. 😅 

I also took the liberty of replacing `atol` with `rtol` here. Without domain knowledge, I don't know whether an absolute difference of 0.01 is big or small here - but a difference of 1% I know we can probably live with, especially for data pulled off a plot.

Sjbrownian replied:
I will definitely keep the `np.testing.assert_allclose` change in mind for future tests. I'm unsure how to tell if `rtol` or `atol` would be better to use in this case. I'll wait for @rocco8773 Erik's thoughts on the matter before committing this change. 
Thank you for the help Dominik!

rocco8773 commented on plasmapy/dispersion/numerical/tests/test_hollweg_.py:
```diff
@@ -0,0 +1,250 @@
+        """
+        Test calculated values based on Figure 2 of Hollweg1999
+        (DOI: https://doi.org/10.1029/1998JA900132) using eqn 3 of Bellan 2012
+        (DOI: https://agupubs.onlinelibrary.wiley.com/doi/10.1029/2012JA017856).
```
```suggestion
        (DOI: https://doi.org/10.1029/2012JA017856).
```

rocco8773 commented on plasmapy/dispersion/numerical/tests/test_hollweg_.py:
```diff
@@ -0,0 +1,250 @@
+    )
+    def test_hollweg1999_vals(self, kwargs, expected):
+        """
+        Test calculated values based on Figure 2 of Hollweg1999
+        (DOI: https://doi.org/10.1029/1998JA900132) using eqn 3 of Bellan 2012
+        (DOI: https://agupubs.onlinelibrary.wiley.com/doi/10.1029/2012JA017856).
+
+        The WebPlotDigitizer software (https://automeris.io/WebPlotDigitizer/)
+        was used to determine the values in Figure 2 of Hollweg1999.
```
```suggestion
        Test calculated values based on Figure 2 of Hollweg1999
        (DOI: https://doi.org/10.1029/1998JA900132) using eqn 3 of
        Bellan 2012
        (DOI: https://doi.org/10.1029/2012JA017856).

        The WebPlotDigitizer software was used to determine the test
        parameters for k, B, and expected omega from Figure 2 of
        Hollweg1999.
        
        - GitHub: https://github.com/ankitrohatgi/WebPlotDigitizer
        - Web Version: https://automeris.io/WebPlotDigitizer/
```

rocco8773 commented on plasmapy/dispersion/numerical/tests/test_hollweg_.py:
```diff
@@ -0,0 +1,250 @@
+        va = pfp.va_(kwargs["B"], kwargs["n_i"], ion=kwargs["ion"]).value
+        wci = pfp.wc_(kwargs["B"], kwargs["ion"]).value
+
+        L = cs / np.abs(wci)
+        kx = (np.sin(kwargs["theta"]) * kwargs["k"]).value
```
```suggestion
```

rocco8773 commented on changelog/1189.feature.rst:
```diff
@@ -0,0 +1,2 @@
+Addition of the `hollweg_.py` module to the `~plasmapy.dispersion.numerical`
+subpackage to numerically solve the dispersion relation using Hollweg's method.
```
```suggestion
Addition of the :file:`hollweg_.py` module to the `~plasmapy.dispersion.numerical`
subpackage to numerically solve the dispersion relation using Hollweg's method,
:cite:p:`hollweg:1999,bellan:2012`.

rocco8773 commented on plasmapy/dispersion/numerical/tests/test_hollweg_.py:
```diff
@@ -0,0 +1,250 @@
+            hollweg(**kwargs)
+
+    @pytest.mark.parametrize(
+        "kwargs, expected",
```
```suggestion
        "kwargs, expected, desired_beta",
```

rocco8773 commented on plasmapy/dispersion/numerical/tests/test_hollweg_.py:
```diff
@@ -0,0 +1,250 @@
+        [
+            (  # beta = 1/20 for kx*L = 0
+                {**_kwargs_hollweg1999, "k": 1e-14 * u.rad / u.m, "B": 6.971e-8 * u.T},
+                1 + 0j,
```
```suggestion
                1 + 0j,
                1/20,
```

rocco8773 commented on plasmapy/dispersion/numerical/tests/test_hollweg_.py:
```diff
@@ -0,0 +1,250 @@
+                    "k": 0.0000439223874624874 * u.rad / u.m,
+                    "B": 6.971e-8 * u.T,
+                },
+                1.4018 + 0j,
```
```suggestion
                1.4018 + 0j,
                1/20,
```

rocco8773 commented on plasmapy/dispersion/numerical/tests/test_hollweg_.py:
```diff
@@ -0,0 +1,250 @@
+            ),
+            (  # beta = 1/2 for kx*L = 0
+                {**_kwargs_hollweg1999, "k": 1e-14 * u.rad / u.m, "B": 2.205e-8 * u.T},
+                1 + 0j,
```
```suggestion
                1 + 0j,
                0.5,
```

rocco8773 commented on plasmapy/dispersion/numerical/tests/test_hollweg_.py:
```diff
@@ -0,0 +1,250 @@
+                    "k": 0.000013893109303101 * u.rad / u.m,
+                    "B": 2.205e-8 * u.T,
+                },
+                1.3536 + 0j,
```
```suggestion
                1.3536 + 0j,
                0.5,
```

rocco8773 commented on plasmapy/dispersion/numerical/tests/test_hollweg_.py:
```diff
@@ -0,0 +1,250 @@
+            ),
+            (  # beta = 2 for kx*L = 0
+                {**_kwargs_hollweg1999, "k": 1e-14 * u.rad / u.m, "B": 1.097e-8 * u.T},
+                1 + 0j,
```
```suggestion
                1 + 0j,
                2,
```

rocco8773 commented on plasmapy/dispersion/numerical/tests/test_hollweg_.py:
```diff
@@ -0,0 +1,250 @@
+                    "k": 0.00000691190063354451 * u.rad / u.m,
+                    "B": 1.097e-8 * u.T,
+                },
+                1.2607 + 0j,
```
```suggestion
                1.2607 + 0j,
                2,
```

rocco8773 commented on plasmapy/dispersion/numerical/tests/test_hollweg_.py:
```diff
@@ -0,0 +1,250 @@
+                    "k": 1e-14 * u.rad / u.m,
+                    "B": 6.97178e-7 * u.T,
+                },
+                1 + 0j,
```
```suggestion
                1 + 0j,
                1/2000,
```

rocco8773 commented on plasmapy/dispersion/numerical/tests/test_hollweg_.py:
```diff
@@ -0,0 +1,250 @@
+                    "k": 0.000439273010336778 * u.rad / u.m,
+                    "B": 6.97178e-7 * u.T,
+                },
+                0.98750 + 0j,
```
```suggestion
                0.98750 + 0j,
                1/2000,
```

rocco8773 commented on plasmapy/dispersion/numerical/tests/test_hollweg_.py:
```diff
@@ -0,0 +1,250 @@
+        """
+        # k values need to be single valued for this test to function correctly
+
+        cs = pfp.cs_(kwargs["T_e"], kwargs["T_i"], kwargs["ion"]).value
+        va = pfp.va_(kwargs["B"], kwargs["n_i"], ion=kwargs["ion"]).value
+        wci = pfp.wc_(kwargs["B"], kwargs["ion"]).value
```
```suggestion
        cs = pfp.cs_(kwargs["T_e"], kwargs["T_i"], kwargs["ion"]).value
        va = pfp.va_(kwargs["B"], kwargs["n_i"], ion=kwargs["ion"]).value

        beta = (cs / va) ** 2
        if not np.isclose(beta, desired_beta, atol=2e-4):
            pytest.fail(
                f"The Holweg 1999 paper requires a 'beta' value of {desired_beta:0.5f} "
                f"and the test parameters yielded {beta:.6f}."
            )
```

rocco8773 commented on plasmapy/dispersion/numerical/tests/test_hollweg_.py:
```diff
@@ -0,0 +1,250 @@
+        w_alfven = (hollweg(**kwargs)["alfven_mode"]).value
+        big_omega = np.abs(w_alfven / (kz * va))
+
+        assert np.isclose(big_omega, expected, atol=1e-2)
```
```suggestion
        assert np.allclose(big_omega, expected, atol=1e-2)
```

rocco8773 commented on plasmapy/dispersion/numerical/tests/test_hollweg_.py:
```diff
@@ -0,0 +1,267 @@
+                0.5,
+            ),
+            (  # beta = 2 for kx*L = 0
+                {**_kwargs_hollweg1999, "k": 1e-14 * u.rad / u.m, "B": 1.097e-8 * u.T},
```
```suggestion
                {**_kwargs_hollweg1999, "k": 1e-14 * u.rad / u.m, "B": 1.10232e-8 * u.T},
```

rocco8773 commented on plasmapy/dispersion/numerical/tests/test_hollweg_.py:
```diff
@@ -0,0 +1,267 @@
+                {
+                    **_kwargs_hollweg1999,
+                    "k": 0.00000691190063354451 * u.rad / u.m,
+                    "B": 1.097e-8 * u.T,
```
```suggestion
                    "B": 1.10232e-8 * u.T,
```

rocco8773 commented on plasmapy/dispersion/numerical/tests/test_hollweg_.py:
```diff
@@ -0,0 +1,267 @@
+                1.2607 + 0j,
+                2,
+            ),
+            (
```
```suggestion
            (  # beta = 1/2000 for kx*L = 0
```

rocco8773 commented on plasmapy/dispersion/numerical/tests/test_hollweg_.py:
```diff
@@ -0,0 +1,267 @@
+            ),
+        ],
+    )
+    def test_hollweg1999_vals(self, kwargs, expected):
```
```suggestion
    def test_hollweg1999_vals(self, kwargs, expected, desired_beta):
```

rocco8773 commented on plasmapy/dispersion/numerical/tests/test_hollweg_.py:
```diff
@@ -0,0 +1,267 @@
+        w_alfven = (hollweg(**kwargs)["alfven_mode"]).value
+        big_omega = np.abs(w_alfven / (kz * va))
+
+        np.testing.assert_allclose(big_omega, expected, rtol=1e-2)
```
```suggestion
        assert np.allclose(big_omega, expected, atol=1e-2)
```

rocco8773 commented on plasmapy/dispersion/numerical/tests/test_hollweg_.py:
```diff
@@ -0,0 +1,267 @@
+
+        The WebPlotDigitizer software was used to determine the test
+        parameters for k, B, and expected omega from Figure 2 of
+        Hollweg1999.
+        
+        - GitHub: https://github.com/ankitrohatgi/WebPlotDigitizer
```
```suggestion
        Hollweg1999.

        - GitHub: https://github.com/ankitrohatgi/WebPlotDigitizer
```

rocco8773 reviewed (approved):
LGTM! Congratulations and thanks for sticking with it!
